OpenGL Optimization work

JeroMiya opened the pull request.

JeroMiya commented:
Note: due to the number of changes in this PR, I'm going to do a clang-format in a separate PR so it doesn't make a bunch of diffs.

russell-taylor commented on osvr/RenderKit/RenderManagerOpenGL.cpp:
```diff
@@ -795,43 +841,19 @@ namespace renderkit {
-		GLint fb;
-		glGetIntegerv(GL_FRAMEBUFFER_BINDING, &fb);
-		m_initialFrameBuffer = static_cast<GLuint>(fb);
-
```
Aren't we worried that this will mess up client code that relies on the frame buffer being the way they left it?  At least one of these store/retrieve calls was put in there in response to an issue where a client was having problems.

russell-taylor replied:
Ahh... looks like this as been handled in another section of the code below.

JeroMiya replied:
Yes, there is now a boolean to control whether to save and restore state: m_storeClientGLState as this has a performance impact. By default state is saved and restored, but the ATW backend turns it off because the ATW thread has its own context, so it's safe to skip the save/restore steps.

russell-taylor replied:
Excellent.

russell-taylor commented on osvr/RenderKit/RenderManagerOpenGL.cpp:
```diff
@@ -795,43 +841,19 @@ namespace renderkit {
 
         // Render to our framebuffer
-        glBindFramebuffer(GL_FRAMEBUFFER, m_frameBuffers.at(GetDisplayUsedByEye(eye)));
+        glBindFramebuffer(GL_FRAMEBUFFER, m_frameBuffers.at(eye));
```
This was in here to make possible the case where the client wants to render both eyes into the same frame buffer.  Is this being handled differently now, or does this change need to be reverted?

russell-taylor replied:
Looks like the framebuffer is being bound below based on whether the eyes are the same.  Do we still need this binding here?

JeroMiya replied:
So, I'm not sure exactly what is going on here because `m_frameBuffers` is created in `constructRenderBuffers` for the `Render` path and it always creates one framebuffer per eye, which seems to indicate that `m_frameBuffers.at(eye)` is the correct way to get the framebuffer. But, the comment in the header for `m_frameBuffers` seems to indicate it is "one per display", but that's not what the code is doing when it creates the frameBuffers. This code here (prior to the change) and the code in the OpenGL/D3D11 interop backend bind to the framebuffers in `m_frameBuffers` uses DISPLAY indices not eye indices, so in the end I'm not sure it's even working?

The code below that binds the frame buffer is binding to the framebuffer for a given display (always one per display) using the toolkit override if available, or else defaults to 0. This is probably technically wrong for multiple displays but usually ends up working because we usually have multiple contexts when using multiple displays and therefor the display framebuffer is typically 0 for a given display's context. We should probably have a more robust solution but that's maybe beyond the scope of this PR.

JeroMiya replied:
I'm going to revert this change as it seems to not be related to the optimization work and we can revisit the Render path another time.

russell-taylor replied:
Good; the GetDisplayUsedByEye() was in there exactly to handle the different cases of one or two eyes per display.  It was tested and worked in all cases when it was put in there.

russell-taylor commented on osvr/RenderKit/RenderManagerOpenGL.cpp:
```diff
@@ -795,43 +841,19 @@ namespace renderkit {
-            return false;
-        }
-
         // Call the display set-up callback for each eye, because they each
```
What happens if the client changes the framebuffer in the rendering code?

JeroMiya replied:
Going to revert this change.

russell-taylor commented on osvr/RenderKit/RenderManagerOpenGL.cpp:
```diff
@@ -934,12 +956,16 @@ namespace renderkit {
-#ifndef OSVR_RM_USE_OPENGLES20
+
         if (VAO) {
+#ifdef OSVR_RM_USE_OPENGLES20
```
Not all implementations of OPENGLES 2.0 have the extension.  In particular, the Ubuntu Tegra implementation is missing it.

russell-taylor replied:
Looks like maybe a m_GLVAOExtensionAvailable is being set that should be used here to guard this call?  Except that won't help with linking failure.

JeroMiya replied:
The GL procs are dynamically loaded at runtime, so there shouldn't be any linking failure. The if(VAO) check is sufficient as a guard here, as VAO is only initialized when m_GLVAOExtensionAvailable is true, are we're testing both successful VAO initialization and VAO extension available in one expression.

russell-taylor replied:
Cool.

russell-taylor commented on osvr/RenderKit/RenderManagerOpenGLATW.h:
```diff
@@ -652,7 +653,7 @@ namespace renderkit {
                                                                 timing.hardwareDisplayInterval.microseconds / 1e6);
                 } else {
-                    m_log->error() << "RenderManagerThread::threadFunc() = couldn't get timing info";
+                    //m_log->error() << "RenderManagerThread::threadFunc() = couldn't get timing info";
```
If this is not desired, probably delete it rather than commenting it out.

JeroMiya replied:
Will do.

JeroMiya commented:
OK @russell-taylor I think I addressed all of your questions/comments. Could you take a second look? Thanks!

russell-taylor commented on osvr/RenderKit/RenderManagerOpenGL.cpp:
```diff
@@ -425,19 +430,23 @@ namespace renderkit {
     RenderManagerOpenGL::~RenderManagerOpenGL() {
         if (m_displayOpen) {
-            for (size_t i = 0; i < m_frameBuffers.size(); i++) {
+            for (size_t i = 0; i < GetNumDisplays(); i++) {
```
Given that we're deleting the elements of the m_frameBuffers vector here, it seems to make more sense to iterate based on its actual size than based on something that may or may not be its size, even if GetNumDisplays() just calls m_frameBuffers.size() now because it will keep the code from drifting.  What's better about using the different method here?

JeroMiya replied:
m_frameBuffers elements aren't deleted in this loop - it just attempts to make each display current via the toolkit and if any of them fail, it bails out early. 

That being said, the loop that does delete elements from m_frameBuffers re-uses a loop iterating over m_colorBuffers. I'll separate out those deletes into a separate loop.

JeroMiya replied:
OK, done.

russell-taylor commented on osvr/RenderKit/RenderManagerOpenGL.cpp:
```diff
@@ -795,20 +841,14 @@ namespace renderkit {
 
         // Render to our framebuffer
-        glBindFramebuffer(GL_FRAMEBUFFER, m_frameBuffers.at(GetDisplayUsedByEye(eye)));
+        glBindFramebuffer(GL_FRAMEBUFFER, GetDisplayUsedByEye(eye));
```
This seems like it is removing a level of indirection that was put in on purpose.  It may work in some cases, but this lookup is not a performance hit so why not put it back the way it was?

JeroMiya replied:
I've already reverted this change.

JeroMiya commented:
OK @russell-taylor I cleaned up the `RenderManagerOpenGL` destructor and separated out the loops so that there is one for each `std::vector` rather than reusing a loop for one and assuming they're the same size. Your other added comment was for a change that I had already reverted.

russell-taylor commented:
Sounds good.  I'm assuming that you have tested it on various OS/graphics configurations; I have not.
